ddtrace/tracer/tracer.go

```diff
@@ -616,6 +624,7 @@ func (t *tracer) sample(span *span) {
 	sampler := t.config.sampler
 	if !sampler.Sample(span) {
 		span.context.trace.drop()
+		span.setSamplingPriority(ext.PriorityAutoReject, samplernames.RuleRate)
```

Comment:
This line is creating an additional allocation [here](https://github.com/DataDog/dd-trace-go/blob/main/ddtrace/tracer/spancontext.go#L323), because we unfortunately use a pointer to a float for the priority 😢 

This needs to be done due to the reasons described in https://github.com/DataDog/dd-trace-go/pull/2067:

> Set the sampling priority of dropped spans by the tracer rate sampler. Previously an unset priority would cause the new partial flushing code to panic as it expected that a sampling decision would have already been made and therefore that `trace.SamplingPriority` would be non-nil.

We could likely get away with changing this line to be:

```
span.context.setSamplingPriority(ext.PriorityAutoReject, samplernames.RuleRate)
```

instead, to avoid setting the sampling priority on every single span, which we probably don't need to do, but the same allocation issue still stands.

Reply:
https://github.com/DataDog/dd-trace-go/pull/2155 removes that pointer, if it's something we end up deeming worthwhile.

Reply:
https://github.com/DataDog/dd-trace-go/pull/2157 I think this is what we should do instead. I know this benchmark *looks* like there's a major regression, but it's actually no where near as bad as it seems, since this allocation and performance hit will only affect one span in an entire trace, not every span.